Do not pull images when running tasks

psavery opened the pull request:
This commit results in tasks running without pulling images.

It currently works for docker, but it probably does not work
for singularity or shifter.

It skips the generation of the image_uri, so, rather than using
the image_uri, we need to create the run commands manually. This
still needs to be done for singularity and shifter.

Depends on: OpenChemistry/mongochemserver#183

psavery committed:
Do not pull images when running tasks with docker

This commit results in tasks running without pulling images.

It currently works for docker, but it probably does not work
for singularity or shifter.

It skips the generation of the image_uri, so, rather than using
the image_uri, we need to create the run commands manually. This
still needs to be done for singularity and shifter.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery commented on taskflows/taskflows/__init__.py:
```diff
@@ -491,26 +459,22 @@ def _create_job(task, input_, cluster, image, run_parameters, container_descript
 
-    if container == 'docker':
-        # In the docker case we need to ensure the image has been pull on this
-        # node, as the images are not shared across the nodes.
```
This looks like something we will need to take into account if we do not automatically pull images.

psavery replied:
@cryos For the work that tries to separate out "docker pull" from the tasks, the above is a potential issue that Chris and I discussed earlier.

I guess part of the reason "docker pull" was originally implemented in the tasks is because, if we are running a task with multiple nodes, each node needs to have its own copy of the image (this is true for singularity as well if the nodes do not share a filesystem).

However, we think we could still move this to some kind of management interface in the client, and if the user chooses to update all docker images, a task could be launched that runs "docker pull" (and the corresponding singularity commands) on every node. Does this sound reasonable?

cryos replied:
This sounds reasonable, and I think it would be a good path forward.

psavery committed:
Add path and extension for singularity sif files

This attempts to add a path and the ".sif" extension for the
singularity sif files. It expects the ".sif" file to already be
residing in the singularity directory before the taskflow is ran.

This is untested.

We will need to add another taskflow soon that will pull images, create
singularity files, etc. when the user asks it to be done.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery committed:
For singularity, ensure dir exists, overwrite sif

For singularity, make the directory on the server, and overwrite the
sif file if it already exists.

Singularity appears to be working with this update.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery commented:
Note to self: before this will be ready for review, I need to make sure the calculation stores the image digest just like the master branch does. This is important for us. Currently, this branch does not store the image digest. It would be nice if it did like so:

```
"image": {
  "digest": "sha256:46ed72cebdfa78ab1ed49d10a3e0e5b3b979971f8b8ef2c1583e1a5c42accd20",
  "imageUri": "openchemistry/psi4@sha256:46ed72cebdfa78ab1ed49d10a3e0e5b3b979971f8b8ef2c1583e1a5c42accd20",
  "repository": "openchemistry/psi4",
  "tag": "latest"
}
```

psavery committed:
Added a taskflow to pull containers

This taskflow will be used by an image manager in mongochemclient.
The user will be able to select the type of container, the
repository, and the tag.

It seems to work for docker just fine. Hopefully it will work for
shifter.

Singularity is not completely working. Although the image appears to
be built, I am not actually sure where it is being placed. There
may be an error at the end of building for an invalid directory
or something.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery committed:
Add imports for both kinds of taskflows

This adds the import for the container pull taskflow
as well.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery committed:
Upload image info to database when one is pulled

This keeps track of which images are in the database, and which aren't.

This commit also adds some checks to ensure the image is in the database
before it is used. This prevents the image from being pulled
automatically when it is ran.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery committed:
Add nicer "image not found" error message

Rather than printing a stack trace in the notebook from the Exception,
just print that the image was not found. This is easier to read.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery committed:
Add a taskflow to register images

This taskflow automatically finds and registers a list of whitelisted
images that are found on the cluster. This will allow the list of images
to be populated the first time the server is started. But currently, the
taskflow isn't being called anywhere.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

cjh1 commented on taskflows/taskflows/container_list_taskflow.py:
```diff
@@ -0,0 +1,250 @@
+        })
+    elif is_demo(cluster):
+        # DEV/DEMO environment options
+        setup_commands = ['source scl_source enable python27']
```
What is this for?

cjh1 commented on taskflows/taskflows/container_list_taskflow.py:
```diff
@@ -0,0 +1,250 @@
+        })
+    elif is_demo(cluster):
+        # DEV/DEMO environment options
+        setup_commands = ['source scl_source enable python27']
```
What is this for?

psavery replied:
I'm not sure. It was copied over from [here](https://github.com/OpenChemistry/openchemistrypy/blob/c63a21910515f060618c484a8f9fbccb88efbfa0/taskflows/taskflows/__init__.py#L211) (the current master branch).

cjh1 replied:
Fair enough!

cjh1 commented on taskflows/taskflows/container_pull_taskflow.py:
```diff
@@ -0,0 +1,279 @@
+    except HttpError as e:
+        # Just ignore the error if the image already exists
+        error_str = 'Image already exists'
+        if error_str not in e.responseText:
```
This seems a little fragile to me. Why not use a status code, for example 409.

psavery replied:
This is now fixed.

psavery committed:
Check for image already exists error with status

Checking via string is a little fragile. Check with the status
code instead.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery commented:
@cjh1 This one is ready for review again. This is probably the biggest PR, and it can go first.

psavery commented:
Actually, this one and OpenChemistry/mongochemserver#183 depend on each other. So I guess they should get merged at the same time...

psavery committed:
Rename "taskBody" to "taskFlowInput"

It is actually input to the taskflow, it is not a task body.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery committed:
Update posting images to include clusterId

This matches what is now set in mongochemserver

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>
